x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java

```diff
@@ -0,0 +1,233 @@
+            // we want to control the refresh of cluster info updates
+            .put(InternalClusterInfoService.INTERNAL_CLUSTER_INFO_UPDATE_INTERVAL_SETTING.getKey(), "60m")
+            .build();
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        Collection<Class<? extends Plugin>> plugins = new ArrayList<>(super.nodePlugins());
+        plugins.add(LocalStateSearchableSnapshots.class);
+        return plugins;
+    }
+
+    @Override
+    protected boolean addMockInternalEngine() {
+        return false;
+    }
+
+    public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception {
+        internalCluster().startMasterOnlyNode();
+        final String dataHotNode = internalCluster().startNode(
+            Settings.builder()
+                .putList(NodeRoleSettings.NODE_ROLES_SETTING.getKey(), DiscoveryNodeRole.DATA_HOT_NODE_ROLE.roleName())
+                .build()
+        );
+
+        final var masterInfoService = (InternalClusterInfoService) internalCluster().getCurrentMasterNodeInstance(ClusterInfoService.class);
+        ClusterInfoServiceUtils.refresh(masterInfoService);
+
+        final int nbIndices = randomIntBetween(2, 5);
+        final Thread[] threads = new Thread[nbIndices];
+        final CountDownLatch latch = new CountDownLatch(nbIndices);
+
+        for (int i = 0; i < nbIndices; i++) {
+            final String index = "index-" + i;
+            var thread = new Thread(() -> {
+                createIndex(
+                    index,
+                    Settings.builder()
+                        .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                        .put(DataTier.TIER_PREFERENCE, DataTier.DATA_HOT)
+                        .put(INDEX_SOFT_DELETES_SETTING.getKey(), true)
+                        .put(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), "0ms")
+                        .put(DataTier.TIER_PREFERENCE_SETTING.getKey(), DataTier.DATA_HOT)
+                        .build()
+                );
+                int nbDocs = 100;
+                try (BackgroundIndexer indexer = new BackgroundIndexer(index, client(), nbDocs)) {
+                    while (true) {
+                        waitForDocs(nbDocs, indexer);
+                        Map<String, Long> storeSize = sizeOfShardsStores(index);
+                        if (storeSize.get(index) > WATERMARK_BYTES) {
+                            break;
+                        }
+                        nbDocs += scaledRandomIntBetween(100, 5_000);
+                    }
+                } catch (Exception e) {
+                    throw new AssertionError(e);
+                } finally {
+                    latch.countDown();
+                }
+            });
+            threads[i] = thread;
+            thread.start();
+        }
+
+        latch.await();
```

Comment:
NIT: I think it is better to add a timeout (just in case) and verify await if finished with true.

```suggestion
        assertThat(latch.await(10, SECONDS), equalTo(true));
```

I think joining the threads would become unnecessary then, unless I miss something

Reply:
I prefer not to add a timeout here: each request already has its own timeout and some CI worker can be very slow. I also don't think joining the threads are an issue?